client/allocrunner/taskrunner/template_hook_test.go

```diff
@@ -135,3 +141,129 @@ func Test_templateHook_Prestart_ConsulWI(t *testing.T) {
 		})
 	}
 }
+
+func Test_templateHook_Prestart_Vault(t *testing.T) {
+	ci.Parallel(t)
+
+	secretsResp := `
+{
+  "data": {
+    "data": {
+      "secret": "secret"
+    },
+    "metadata": {
+      "created_time": "2023-10-18T15:58:29.65137Z",
+      "custom_metadata": null,
+      "deletion_time": "",
+      "destroyed": false,
+      "version": 1
+    }
+  }
+}`
+
+	// Start test server to simulate Vault cluster responses.
+	reqCh := make(chan any)
+	defaultVaultServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		reqCh <- struct{}{}
+		fmt.Fprintln(w, secretsResp)
+	}))
+	t.Cleanup(defaultVaultServer.Close)
+
+	// Setup client with Vault config.
+	clientConfig := config.DefaultConfig()
+	clientConfig.TemplateConfig.DisableSandbox = true
+	clientConfig.VaultConfigs = map[string]*structsc.VaultConfig{
+		structs.VaultDefaultCluster: {
+			Name:    structs.VaultDefaultCluster,
+			Enabled: pointer.Of(true),
+			Addr:    defaultVaultServer.URL,
+		},
+	}
+
+	testCases := []struct {
+		name            string
+		vault           *structs.Vault
+		expectedCluster string
+	}{
+		{
+			name: "use default cluster",
+			vault: &structs.Vault{
+				Cluster: structs.VaultDefaultCluster,
+			},
+			expectedCluster: structs.VaultDefaultCluster,
+		},
+		{
+			name:            "use default cluster if no vault block is provided",
+			vault:           nil,
+			expectedCluster: structs.VaultDefaultCluster,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// Setup alloc and task to connect to Vault cluster.
+			alloc := mock.MinAlloc()
+			task := alloc.Job.TaskGroups[0].Tasks[0]
+			task.Vault = tc.vault
+
+			// Setup template hook.
+			taskDir := t.TempDir()
+			hookConfig := &templateHookConfig{
+				alloc:        alloc,
+				logger:       testlog.HCLogger(t),
+				lifecycle:    trtesting.NewMockTaskHooks(),
+				events:       &trtesting.MockEmitter{},
+				clientConfig: clientConfig,
+				envBuilder:   taskenv.NewBuilder(mock.Node(), alloc, task, clientConfig.Region),
+				templates: []*structs.Template{
+					{
+						EmbeddedTmpl: `{{with secret "secret/data/test"}}{{.Data.data.secret}}{{end}}`,
+						ChangeMode:   structs.TemplateChangeModeNoop,
+						DestPath:     path.Join(taskDir, "out.txt"),
+					},
+				},
+			}
+			hook := newTemplateHook(hookConfig)
+
+			// Start template hook with a timeout context to ensure it exists.
+			req := &interfaces.TaskPrestartRequest{
+				Task:    task,
+				TaskDir: &allocdir.TaskDir{Dir: taskDir},
+			}
+
+			ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+			t.Cleanup(cancel)
+
+			// Start in a goroutine because Prestart() blocks until first
+			// render.
+			hookErrCh := make(chan error)
+			go func() {
+				err := hook.Prestart(ctx, req, nil)
+				hookErrCh <- err
+			}()
+
+			var gotRequest bool
+		LOOP:
```

Comment:
Because the send on `reqCh` is blocking, at first I thought you could probably select over `reqCh` and `ctx` without this loop and check the error afterwards, and avoid the label/break/goto. But this is a nicely elegant way of avoiding blocking 3 whole seconds for the context timeout on error. :+1:

Reply:
Yeah, this ensure we exit as soon as `Prestart()` finishes. I think there's a theoretical issue where a test case reads a request from another test case, because `reqCh` is shared while each test case has its own consumer. But since this is a test, and each test case doesn't run in parallel, I think it's fine.